tls/extensions/s2n_client_key_share.c

```diff
@@ -319,70 +437,42 @@ static int s2n_client_key_share_recv(struct s2n_connection *conn, struct s2n_stu
     GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
     notnull_check(ecc_pref);
 
+    const struct s2n_kem_preferences *kem_pref = NULL;
+    GUARD(s2n_connection_get_kem_preferences(conn, &kem_pref));
+    notnull_check(kem_pref);
+
     uint16_t key_shares_size;
     GUARD(s2n_stuffer_read_uint16(extension, &key_shares_size));
-    S2N_ERROR_IF(s2n_stuffer_data_available(extension) < key_shares_size, S2N_ERR_BAD_MESSAGE);
+    ENSURE_POSIX(s2n_stuffer_data_available(extension) >= key_shares_size, S2N_ERR_BAD_MESSAGE);
 
-    const struct s2n_ecc_named_curve *supported_curve;
-    struct s2n_blob point_blob;
     uint16_t named_group, share_size;
-    uint32_t supported_curve_index;
-
-    /* Whether a match was found */
-    uint8_t match = 0;
-
+    bool match_found = false;
     /* bytes_processed is declared as a uint32_t to avoid integer overflow in later calculations */
     uint32_t bytes_processed = 0;
 
     while (bytes_processed < key_shares_size) {
         GUARD(s2n_stuffer_read_uint16(extension, &named_group));
         GUARD(s2n_stuffer_read_uint16(extension, &share_size));
 
-        S2N_ERROR_IF(s2n_stuffer_data_available(extension) < share_size, S2N_ERR_BAD_MESSAGE);
+        ENSURE_POSIX(s2n_stuffer_data_available(extension) >= share_size, S2N_ERR_BAD_MESSAGE);
         bytes_processed += share_size + S2N_SIZE_OF_NAMED_GROUP + S2N_SIZE_OF_KEY_SHARE_SIZE;
 
-        supported_curve = NULL;
-        for (size_t i = 0; i < ecc_pref->count; i++) {
-            if (named_group == ecc_pref->ecc_curves[i]->iana_id) {
-                supported_curve_index = i;
-                supported_curve = ecc_pref->ecc_curves[i];
-                break;
-            }
-        }
-
-        /* Ignore unsupported curves */
-        if (!supported_curve) {
-            GUARD(s2n_stuffer_skip_read(extension, share_size));
-            continue;
-        }
-
-        /* Ignore curves that we've already received material for */
-        if (conn->secure.client_ecc_evp_params[supported_curve_index].negotiated_curve) {
-            GUARD(s2n_stuffer_skip_read(extension, share_size));
-            continue;
-        }
-
-        /* Ignore curves with unexpected share sizes */
-        if (supported_curve->share_size != share_size) {
-            GUARD(s2n_stuffer_skip_read(extension, share_size));
-            continue;
-        }
-
-        GUARD(s2n_ecc_evp_read_params_point(extension, share_size, &point_blob));
+        DEFER_CLEANUP(struct s2n_stuffer key_share = { 0 }, s2n_stuffer_free);
+        GUARD(s2n_stuffer_alloc(&key_share, share_size));
+        GUARD(s2n_stuffer_copy(extension, &key_share, share_size));
```

Comment:
Do we need to `alloc` and `free` here? I think we can create a temp blob from a `s2n_stuffer_raw_read()` and initialize a new stuffer with the temp blob without allocating or copying memory.

Reply:
Good call. Fixed.